src/main/java/codeu/controller/AdminServlet.java

```diff
@@ -11,32 +14,44 @@
 /** Servlet class responsible for the admin page. */
 public class AdminServlet extends HttpServlet {
 
-    private static final String ADMIN_MESSAGE = "admin_message";
+    private static final String ADMIN_MESSAGE = "adminMessage";
+    private static final String NUM_USERS = "numUsers";
+    private static final String NUM_MESSAGES = "numMessages";
+    private static final String NUM_CONVERSATIONS = "numConversations";
+    private static final String ERROR = "error";
 
     /**
      * This function fires when a user navigates to the admin page. It displays a message on the page, which differs
      * depending on if you are logged in, and whether or not you are an admin.
+     * The page contains information pertaining to administration if you are an admin.
      */
     @Override
     protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
-
         String user = (String) request.getSession().getAttribute("user");
 
-        /* Note: the page should always have "Welcome to the admin page!" because I later edited the index.jsp page to
-        only show the admin link if you are currently logged in as an admin. I (Justin) will refactor this code for
-        the MVP.
-        */
         if (user != null) {
             if (AdminHelper.isAdmin(user)) {
                 request.getSession().setAttribute(ADMIN_MESSAGE, "Hello " + user + "! Welcome to the admin page!");
             } else {
                 request.getSession().setAttribute(ADMIN_MESSAGE, "You are not an admin!");
+                request.getRequestDispatcher("/WEB-INF/view/admin.jsp").forward(request, response);
+                return;
             }
         } else {
-            response.sendRedirect("/login");
+            request.setAttribute(ERROR, "You must log in before you can view the admin page!");
+            request.getRequestDispatcher("/WEB-INF/view/login.jsp").forward(request, response);
```

Comment:
Maybe it's better to redirect them here rather than render the login page because there may be some state transition that we are not respecting or assumption on the path that we are breaking in the login page. If needed you could add a "post_login_redirect" parameter to the login page and set that here to "/admin"

Reply:
If I redirect, is there any good way for me to set the login page's error message attribute? I changed it to render the login page because I couldn't get the error message to show on the login page by just redirecting.

Also, by the post_login_redirect parameter you mean that I could use that in doPost in the loginServlet to figure out which page to redirect back to after a successful login?

Reply:
1. You could pass that as a URL parameter and use that in rendering the template ("/login?error_message=notloggedin").

2. Yes. You'd also have to make sure the parameter is propagated. It's a completely optional suggestion.

Reply:
Thanks for these changes/suggestions, just pushed them